### DISCUSSION TITLE
Short <LASTLOGGEDUSER>

### DISCUSSION THREAD

**Discussion by Krovyakov-Ivan** [glpi-project/glpi-agent#748] [Q&A]
Good afternoon. 
In the agent, the glpi-inventory command shows the abbreviated user login in the ```<LASTLOGGEDUSER>``` field, for example ```<LASTLOGGEDUSER>krovyako</LASTLOGGEDUSER>```
But you should 
```<LASTLOGGEDUSER>krovyakoviv</LASTLOGGEDUSER>```
Is it possible that this is a bug? 
Agent Version 1.10

**Comment by g-bougard** [comment]
Hi @Krovyakov-Ivan 

I moved your question to Q&A.

This field is surely extracted from registry. If this is indeed a bug, to me this is an OS bug. Agent is not intended to cut found values. Did you have the same problem with older version of the agent ?

**Reply by Krovyakov-Ivan** [reply]
I tested another theory.
When I run ```glpi-agent --full```, the full contact information is displayed on the server in Historical, that is, there is ```Alternate Username = krovyakoviv```
But if I run ```glpi-agent```, then ```Alternate Username = krovyako```
(I can't attach a screenshot)
```glpi-inventory``` is always shown ```<LASTLOGGEDUSER>krovyako</LASTLOGGEDUSER>```

**Reply by g-bougard** [reply]
Since GLPI 10.0.16, users inventory support has been re-enabled during partial inventory. This is important in the case the user is really changing.

But hardware section still contains the LASTLOGGEDUSER, even if USERS section doesn't change. So I think we should now clean up this field in glpi-agent in that context.

Let me know one point, you didn't tell which operating system you're using. I now know the problem can occur on linux as lastloggeduser is extracted from `last` command output and this command keeps username in a size limited field. Can you confirm which OS you're using ?

**Reply by g-bougard** [reply]
Anyway, I just push a patch to glpi-agent here: 7b8a5f0b503a410f6bc1383ea9ed30305afb1842

It should fix your issue. And it will be included in next nightly build so you can test and validate it tomorrow.

**Reply by Krovyakov-Ivan** [reply]
> Since GLPI 10.0.16, users inventory support has been re-enabled during partial inventory. This is important in the case the user is really changing.
> 
> But hardware section still contains the LASTLOGGEDUSER, even if USERS section doesn't change. So I think we should now clean up this field in glpi-agent in that context.
> 
> Let me know one point, you didn't tell which operating system you're using. I now know the problem can occur on linux as lastloggeduser is extracted from `last` command output and this command keeps username in a size limited field. Can you confirm which OS you're using ?

Linux, you're right

**Reply by Krovyakov-Ivan** [reply]
> Anyway, I just push a patch to glpi-agent here: 7b8a5f0b503a410f6bc1383ea9ed30305afb1842
> 
> It should fix your issue. And it will be included in next nightly build so you can test and validate it tomorrow.

So tomorrow I can download the agent's nightly build? By what link?

**Reply by g-bougard** [reply]
Nightly builds are available from that page: https://nightly.glpi-project.org/glpi-agent/

Indeed, you still can test development build from here: https://github.com/glpi-project/glpi-agent/actions/runs/10597984775

**Reply by Krovyakov-Ivan** [reply]
> Nightly builds are available from that page: https://nightly.glpi-project.org/glpi-agent/
> 
> Indeed, you still can test development build from here: https://github.com/glpi-project/glpi-agent/actions/runs/10597984775

I have installed a nightly build. Now, when ```glpi-agent``` and ```glpi-agent --full``` are executed, the full login is displayed on the server, ```Alternate Username = krovyakoviv.```
Thank you for responding so quickly!!!

**Comment by topinet** [comment]
I have same problem, Ubuntu computers get the "Alternate username" trimmed to 8 characters.
The problem is with "last" command used to get lastLoggedUser, which must use the "-w" flag to get full names (not trimmed)
Adding this flag to line 153 of lib/GLPI/Agent/Task/Inventory/Generic/Users.pm solves the problem.

I've raised a bug report for this:
https://github.com/glpi-project/glpi-agent/issues/757

**Comment by Krovyakov-Ivan** [comment]
> I have same problem, Ubuntu computers get the "Alternate username" trimmed to 8 characters.
> The problem is with "last" command used to get lastLoggedUser, which must use the "-w" flag to get full names (not trimmed)
> Adding this flag to line 153 of lib/GLPI/Agent/Task/Inventory/Generic/Users.pm solves the problem.
> 
> I've raised a bug report for this:
> https://github.com/glpi-project/glpi-agent/issues/757

Good afternoon. We are waiting for the official update of the agent.

**Comment by Krovyakov-Ivan** [comment]
Thanks!!! It works great

**Comment by szelga** [comment]
@g-bougard Could there be a quick and dirty solution for domain computers under Linux? AFAIK, `who` and `last` commands trim username to 33 characters, which could be a problem for domain usernames.

**Reply by g-bougard** [reply]
Hi @szelga 

this was the purpose of the [related commit](https://github.com/glpi-project/glpi-agent/commit/53829d3b7c71a0ff618e9f530c9f0d9b446f9514), it makes glpi-agent using `-w` option for last command. For `who`, can you share in a new discussion some outputs where expected output is truncated ?

**Reply by szelga** [reply]
provided that long AD domain name is `domain.company.name`, short is `COMPANY`, and username is `domain.username`, Windows agents provide output like this:

```
    <USERS>
      <DOMAIN>COMPANY</DOMAIN>
      <LOGIN>domain.username</LOGIN>
    </USERS>
```

whereas Linux agents (joined into the AD domain via sssd) provide output like this:

```
   <USERS>
      <LOGIN>domain.username@domain.company.n</LOGIN>
   </USERS>
```

(note it being truncated to 32 characters).

not that I **expected** Linux agents to provide output like Windows ones, but that would be nice, because for Windows machines GLPI matches agent's output with LDAP users w/o any hassle, which is not the case for Linux machines (I have an old GLPI installation, so your results may vary).

I've tried various keys for `who` and `last` to get UID instead of login (and UID could later be converted to a proper username), but to no avail.

(sorry if this is offtopic.)

**Reply by szelga** [reply]
OK, I may have an idea on `who`.

```
# getting process IDs
$ who --users | awk '{print $6}'
4597
15509
17307
32150
49920
# getting UIDs
$ ps -o user:128 -p 4597,15509,17307,32150,49920 | tail -n +2 | sort | uniq
991227614
991226519
# getting full name(s)
$ id -un 991227614 991226519
domain.username@domain.company.name
another.bloody.user@domain.company.name
```

and then split the username and (if SAMBA is installed, and `smb.conf` is configured properly, it should be extracted via `testparm -sl --parameter-name=workgroup` command) substitute domain name for workgroup name (as Windows agents seem to do).

I still have no idea, what to do about `last`, though (**UPD**: having read about `/var/log/lastlog` format, it seems that it truncates username to 32 chars by design, nothing to do about that except for questionable hacks).

**Reply by g-bougard** [reply]
Hi @szelga 

this is interesting, but please open a new "Ideas" discussion to not disturb people who were there a year ago. Name your new discussion with something like "Update users inventory on linux domain computers". You can just copy what you still shared. But can you also share how "domain computers" are setup ? The solution may be to detect commands provided by the solution and get rid of who & last if we have a better replacement.

I'm locking this discussion and we will continue in your new discussion.
